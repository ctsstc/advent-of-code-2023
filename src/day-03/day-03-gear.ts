import {
  Day03Collection,
  Day03Number,
  day03NumberNextToSymbol,
  Day03Symbol,
} from "./day-03-utilities";

export type Day03ValidGearWithNumbers = {
  gear: Day03Symbol;
  gearRatio: number;
};

export function day03ValidGearsWithNumbers(
  collection: Day03Collection
): Day03ValidGearWithNumbers[] {
  const gears = getGears(collection);
  const validGears = gears.filter((gear) => gear.numbers.length === 2);

  return validGears.map((gear) => {
    const firstGearValue = gear.numbers[0].value;
    const secondGearValue = gear.numbers[1].value;
    const gearRatio = firstGearValue * secondGearValue;

    return {
      gear,
      gearRatio,
    };
  });
}

type Day03Gear = Day03Symbol & {
  numbers: Day03Number[];
};

const Day03GearSymbol = "*";

function getGears(collection: Day03Collection): Day03Gear[] {
  const gears = collection.symbols.filter(
    (symbol) => symbol.value === Day03GearSymbol
  );

  return gears.map((gear) => {
    return {
      ...gear,
      numbers: getNumbersForGear(gear, collection.numbers),
    };
  });
}

function getNumbersForGear(
  symbol: Day03Symbol,
  numbers: Day03Number[]
): Day03Number[] {
  return numbers.filter((number) => day03NumberNextToSymbol(number, symbol));
}
