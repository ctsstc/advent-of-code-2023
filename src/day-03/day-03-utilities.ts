type Day03Coordinate = {
  x: number;
  y: number;
};

export type Day03Number = Day03Coordinate & {
  value: number;
  length: number;
};

export type Day03Symbol = Day03Coordinate;

type Day03Collection = {
  numbers: Day03Number[];
  symbols: Day03Symbol[];
};

export function day03ParseLine(line: string, row: number): Day03Collection {
  const matcher = /\d+|[^.\d]/g;
  const matches = [...line.matchAll(matcher)];

  return matches.reduce(
    (collection, match) => {
      const valueString = match[0];
      const value = parseInt(valueString);
      const isSymbol = isNaN(value);

      if (match.index === undefined) throw new Error("Match missing index?!");

      const coordinate: Day03Coordinate = {
        x: match.index,
        y: row,
      };

      if (isSymbol) {
        collection.symbols.push({
          ...coordinate,
        });
      } else {
        collection.numbers.push({
          ...coordinate,
          value,
          length: valueString.length,
        });
      }

      return collection;
    },
    {
      numbers: [] as Day03Number[],
      symbols: [] as Day03Symbol[],
    }
  );
}

export function day03ParseLines(lines: string[]): Day03Collection {
  return lines.reduce(
    (collection, line, column) => {
      const { numbers, symbols } = day03ParseLine(line, column);

      collection.numbers.push(...numbers);
      collection.symbols.push(...symbols);

      return collection;
    },
    {
      numbers: [] as Day03Number[],
      symbols: [] as Day03Symbol[],
    }
  );
}

export function day03ValidNumbers(collection: Day03Collection): Day03Number[] {
  return collection.numbers.filter((number) => {
    return collection.symbols.some((symbol) => {
      // determine if the symbol is within the range of the number
      // a range is the start position and the end
      // a symbol must be either on the same line or within 1 line
      //  then the x must be within the number x - 1, through number x + length + 1

      const xStartRange = number.x - 1;
      const xEndRange = number.x + number.length + 1;
      const yStartRange = number.y - 1;
      const yEndRange = number.y + 1;

      const withinXRange = symbol.x >= xStartRange && symbol.x <= xEndRange;
      const withinYRange = symbol.y >= yStartRange && symbol.y <= yEndRange;

      return withinXRange && withinYRange;
    });
  });
}
