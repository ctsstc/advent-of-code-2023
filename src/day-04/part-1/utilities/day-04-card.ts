export type Day04Card = {
  winningNumbers: number[];
  myNumbers: number[];
};

export function day04ParseCard(line: string): Day04Card {
  const cardSplitString = /\:\s+/;
  const cardNumbersSplitString = /\s+\|\s+/;

  const [_cardId, numbers] = line.split(cardSplitString);
  const [winningNumbersString, myNumbersString] = numbers.split(
    cardNumbersSplitString
  );

  const winningNumbers = numbersStringToNumbers(winningNumbersString);
  const myNumbers = numbersStringToNumbers(myNumbersString);

  return {
    winningNumbers,
    myNumbers,
  };
}

function numbersStringToNumbers(numbersString: string): number[] {
  const numbersSplitString = /\s+/g;

  return numbersString
    .split(numbersSplitString)
    .map((numberString) => parseInt(numberString));
}

export function day04ScoreCard(card: Day04Card): number {
  const unionLength = cardUnion(card).length;

  return Math.pow(2, unionLength - 1);
}

function cardUnion(card: Day04Card): number[] {
  return card.winningNumbers.filter((winningNumber) =>
    card.myNumbers.includes(winningNumber)
  );
}
