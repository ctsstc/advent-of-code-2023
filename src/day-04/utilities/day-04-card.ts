export type Day04Card = {
  winningNumbers: number[];
  myNumbers: number[];
};

export function day04ParseCard(line: string): Day04Card {
  const cardSplitString = /\:\s+/;
  const cardNumbersSplitString = /\s+\|\s+/;

  const [_cardId, numbers] = line.split(cardSplitString);
  const [winningNumbersString, myNumbersString] = numbers.split(
    cardNumbersSplitString
  );

  const winningNumbers = numbersStringToNumbers(winningNumbersString);
  const myNumbers = numbersStringToNumbers(myNumbersString);

  return {
    winningNumbers,
    myNumbers,
  };
}

function numbersStringToNumbers(numbersString: string): number[] {
  const numbersSplitString = /\s+/g;

  return numbersString
    .split(numbersSplitString)
    .map((numberString) => parseInt(numberString));
}

export type Day04CardScore = {
  amount: number;
  length: number;
};

export function day04ScoreCard(card: Day04Card): Day04CardScore {
  const unionLength = cardUnion(card).length;

  if (unionLength === 0)
    return {
      amount: 0,
      length: 0,
    };

  const amount = Math.pow(2, unionLength - 1);

  return {
    amount,
    length: unionLength,
  };
}

function cardUnion(card: Day04Card): number[] {
  return card.winningNumbers.filter((winningNumber) =>
    card.myNumbers.includes(winningNumber)
  );
}
