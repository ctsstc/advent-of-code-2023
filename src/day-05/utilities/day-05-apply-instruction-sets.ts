import { Day05InstructionSet } from "./day-05-instruction-sets";

export function day05ApplyInstructionSets(
  seeds: number[],
  instructionSets: Day05InstructionSet[]
): number[] {
  return seeds.map((seed) => {
    for (const { instructions } of instructionSets) {
      for (const instruction of instructions) {
        const lowerBound = instruction.source;
        const upperBound = instruction.source + instruction.range;
        const inBounds = seed >= lowerBound && seed <= upperBound;
        const difference = seed - lowerBound;

        if (inBounds) {
          seed = instruction.destination + difference;

          // stop going through instructions in this set and move onto the next set
          break;
        }
      }
    }

    return seed;
  });
}
