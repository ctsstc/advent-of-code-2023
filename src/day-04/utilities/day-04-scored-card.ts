import { Day04Card, day04ScoreCard } from "./day-04-card";

export type Day04ScoredCard = Day04Card & {
  score: number;
};

export function day04ScoredCardsFromCards(
  cards: Day04Card[]
): Day04ScoredCard[] {
  return cards.map((card) => {
    return {
      ...card,
      score: day04ScoreCard(card),
    };
  });
}
