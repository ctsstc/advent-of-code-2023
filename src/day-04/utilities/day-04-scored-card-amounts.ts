import { Day04ScoredCard } from "./day-04-scored-card";

export function day04ScoredCardAmounts(
  scoredCards: Day04ScoredCard[]
): Map<number, bigint> {
  // Initial card amounts all start at 1
  const cardAmounts = new Map(
    scoredCards.map((_scoredCard, idx) => [idx, BigInt(1)])
  );

  for (let i = 0; i < scoredCards.length; i++) {
    const scoreLength = scoredCards[i].score.length;
    const multiplier = cardAmounts.get(i);

    if (multiplier === undefined) throw new Error("Issue getting multiplier");

    for (let scoreIdx = 1; scoreIdx <= scoreLength; scoreIdx++) {
      const cardAmountsOffset = i + scoreIdx;
      const currentCardAmount = cardAmounts.get(cardAmountsOffset);

      if (currentCardAmount !== undefined) {
        cardAmounts.set(cardAmountsOffset, currentCardAmount + multiplier);
      }
    }
  }

  return cardAmounts;
}
