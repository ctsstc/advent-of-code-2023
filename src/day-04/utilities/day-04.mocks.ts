import { getLines } from "../../utilities/files";
import { day04ParseCard } from "./day-04-card";
import { day04ScoredCardsFromCards } from "./day-04-scored-card";

export const mockDay04CardsExample = import.meta.resolveSync(
  "../day-04-example.txt"
);
export const mockDay04CardsLines = await getLines(mockDay04CardsExample);
export const mockDay04Cards = mockDay04CardsLines.map((cardLine) =>
  day04ParseCard(cardLine)
);
export const mockDay04ScoredCards = day04ScoredCardsFromCards(mockDay04Cards);
