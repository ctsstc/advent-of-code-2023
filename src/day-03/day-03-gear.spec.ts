import { describe, expect, it } from "bun:test";
import { getLines } from "../utilities/files";
import { day03ValidGearsWithNumbers } from "./day-03-gear";
import { day03ParseLines } from "./day-03-utilities";

describe("Day 03 > Gear", async () => {
  const example = import.meta.resolveSync("./day-03-example.txt");
  const exampleLines = await getLines(example);
  const collection = day03ParseLines(exampleLines);
  const validGears = day03ValidGearsWithNumbers(collection);

  it("returns valid gears and their numbers", () => {
    expect(validGears).toMatchSnapshot();
  });
});
