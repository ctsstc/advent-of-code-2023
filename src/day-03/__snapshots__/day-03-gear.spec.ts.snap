// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`Day 03 > Gear returns valid gears and their numbers 1`] = `
[
  {
    "gear": {
      "numbers": [
        {
          "length": 3,
          "value": 467,
          "x": 0,
          "y": 0,
        },
        {
          "length": 2,
          "value": 35,
          "x": 2,
          "y": 2,
        },
      ],
      "value": "*",
      "x": 3,
      "y": 1,
    },
    "gearRatio": 16345,
  },
  {
    "gear": {
      "numbers": [
        {
          "length": 3,
          "value": 755,
          "x": 6,
          "y": 7,
        },
        {
          "length": 3,
          "value": 598,
          "x": 5,
          "y": 9,
        },
      ],
      "value": "*",
      "x": 5,
      "y": 8,
    },
    "gearRatio": 451490,
  },
]
`;
